### PR TITLE
Python: Implement Hive create and load table

### PR DESCRIPTION
**Pull request by Fokko**

### CODE REVIEW

File: python/pyiceberg/catalog/hive.py [hunk 1]
```diff
@@ -159,16 +156,56 @@ def _construct_parameters(metadata_location: str, previous_metadata_location: Op
...
+    def list(self, list_type: ListType, element_result: str) -> str:
+        return f"array<{element_result}>"
+
+    def map(self, map_type: MapType, key_result: str, value_result: str) -> str:
+        # Key has to be primitive for Hive
+        return f"map<{key_result}, {value_result}>"
+
+    def primitive(self, primitive: PrimitiveType) -> str:
+        if isinstance(primitive, DecimalType):
+            return f"DECIMAL({primitive.precision}, {primitive.scale})"
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:204]: Why is decimal upper case but the others aren't?

File: python/pyiceberg/catalog/hive.py [hunk 2]
```diff
@@ -159,16 +156,56 @@ def _construct_parameters(metadata_location: str, previous_metadata_location: Op
...
+    TimeType: "string",
+    TimestampType: "timestamp",
+    StringType: "string",
+    UUIDType: "string",
+    BinaryType: "binary",
+    FixedType: "binary",
+}
+
+
+class HiveSchemaConstructor(SchemaVisitor[str]):
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:185]: What about naming this `HiveSchemaConverter` or `SchemaToHiveConverter`? I think "converter" describes what is happening here more closely than "constructor" because the schema is based on an Iceberg schema.

File: python/pyiceberg/catalog/hive.py [hunk 3]
```diff
@@ -159,16 +156,56 @@ def _construct_parameters(metadata_location: str, previous_metadata_location: Op
...
-        elif key == "location":
+        elif key == LOCATION:
             database.locationUri = value
         else:
             params[key] = value
     database.parameters = params
     return database
 
 
+hive_primitive_types = {
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:169]: Since this is a constant, should the variable name be all upper case?

File: python/pyiceberg/catalog/hive.py [hunk 4]
```diff
@@ -159,16 +156,56 @@ def _construct_parameters(metadata_location: str, previous_metadata_location: Op
...
+
+class HiveSchemaConstructor(SchemaVisitor[str]):
+    def schema(self, schema: Schema, struct_result: str) -> str:
+        return struct_result
+
+    def struct(self, struct: StructType, field_results: List[str]) -> str:
+        return f"struct<{', '.join(field_results)}>"
+
+    def field(self, field: NestedField, field_result: str) -> str:
+        return f"{field.name}: {field_result}"
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:193]: I'm not sure that there can be a space here. The Java converter doesn't have a space and I seem to remember Hive expecting a certain format in some cases.

File: python/pyiceberg/catalog/hive.py [hunk 5]
```diff
@@ -197,17 +234,41 @@ def __init__(self, name: str, properties: Properties, uri: str):
         super().__init__(name, properties)
         self._client = _HiveClient(uri)
 
-    def _convert_hive_into_iceberg(self, table: HiveTable) -> Table:
-        # Requires reading the manifest, will implement this in another PR
-        # Check the table type
-        return Table(identifier=(table.dbName, table.tableName))
+    def _convert_hive_into_iceberg(self, table: HiveTable, metadata_location: Optional[str] = None) -> Table:
+        properties: Dict[str, str] = table.parameters
+        if TABLE_TYPE not in properties or properties[TABLE_TYPE] != ICEBERG:
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:239]: Can you make the table type check case insensitive? We've had problems with that in the past.

File: python/pyiceberg/catalog/hive.py [hunk 6]
```diff
@@ -197,17 +234,41 @@ def __init__(self, name: str, properties: Properties, uri: str):
...
+            raise NoSuchTableError(f"Table is not an Iceberg table: {table.dbName}.{table.tableName}")
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:240]: Should we add "missing table_type" to this?

File: python/pyiceberg/catalog/hive.py [hunk 7]
```diff
@@ -197,17 +234,41 @@ def __init__(self, name: str, properties: Properties, uri: str):
...
+
+        if not metadata_location:
+            if prop_metadata_location := properties.get(METADATA_LOCATION):
+                metadata_location = prop_metadata_location
+            else:
+                raise NoSuchTableError("Metadata location not found")
+
+        file = PyArrowFile(metadata_location)
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:248]: File implementations should not be used directly. Instead, this should have a `FileIO` instance passed in to use for reading:

```
file = io.newInputFile(metadata_location)
```

File: python/pyiceberg/catalog/hive.py [hunk 8]
```diff
@@ -197,17 +234,41 @@ def __init__(self, name: str, properties: Properties, uri: str):
...
+        metadata = FromInputFile.table_metadata(file)
+        return Table(identifier=(table.dbName, table.tableName), metadata=metadata, metadata_location=metadata_location)
+
+    def _write_metadata(self, metadata: TableMetadataV2, metadata_path: str):
+        output_file = PyArrowFile(metadata_path)
+        # 😢 https://github.com/apache/iceberg/pull/5439
+        os.makedirs(f"{metadata.location}metadata".lstrip("file:"), exist_ok=True)
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:255]: `OutputFile.create` is responsible for handling this.

File: python/pyiceberg/catalog/hive.py [hunk 9]
```diff
@@ -197,17 +234,41 @@ def __init__(self, name: str, properties: Properties, uri: str):
...
+        ToOutputFile.table_metadata(metadata, output_file)
+
+    def _resolve_table_location(self, location: Optional[str], database_name: str, table_name: str):
+        if not location:
+            database_properties = self.load_namespace_properties(database_name)
+            if database_location := database_properties.get(LOCATION):
+                database_location = database_location.rstrip("/")
+                return f"{database_location}/{table_name}/"
+        raise ValueError("Cannot determine location from warehouse, please provide an explicit location")
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:264]: This should respect the `warehouse` property used to construct the catalog rather than failing with `ValueError`

File: python/pyiceberg/catalog/hive.py [hunk 10]
```diff
@@ -229,24 +290,43 @@ def create_table(
             ValueError: If the identifier is invalid
         """
         database_name, table_name = self.identifier_to_database_and_table(identifier)
-        current_time_millis = int(time.time())
+        seconds_since_epoch = int(time.time())
+
+        location = self._resolve_table_location(location, database_name, table_name)
+
+        metadata_location = f"{location}metadata/{uuid.uuid4()}.metadata.json"
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:297]: Paths should always be stored without a trailing `/`, so constructing paths must always add `/` as a separator.

File: python/pyiceberg/catalog/hive.py [hunk 11]
```diff
@@ -229,24 +290,43 @@ def create_table(
...
+        metadata = TableMetadataV2(
+            location=location,
+            schemas=[schema],
+            current_schema_id=schema.schema_id,
+            partition_specs=[partition_spec],
+            default_spec_id=0,
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:303]: I think it's safer to use `partition_spec.spec_id`.

File: python/pyiceberg/catalog/hive.py [hunk 12]
```diff
@@ -229,24 +290,43 @@ def create_table(
...
         database_name, table_name = self.identifier_to_database_and_table(identifier)
-        current_time_millis = int(time.time())
+        seconds_since_epoch = int(time.time())
+
+        location = self._resolve_table_location(location, database_name, table_name)
+
+        metadata_location = f"{location}metadata/{uuid.uuid4()}.metadata.json"
+        metadata = TableMetadataV2(
+            location=location,
+            schemas=[schema],
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:300]: The schema field IDs and partition field IDs need to be reassigned to ensure that they are consistent because we don't trust that users pass them in correctly. Passing them directly is okay for this PR, but we should not release until we have reassignment done.

File: python/pyiceberg/catalog/hive.py [hunk 13]
```diff
@@ -229,24 +290,43 @@ def create_table(
...
+            current_schema_id=schema.schema_id,
+            partition_specs=[partition_spec],
+            default_spec_id=0,
+            sort_orders=[sort_order],
+            default_sort_order_id=0,
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:305]: Sort order ID 0 is reserved for the unsorted order. This can't always assume that the order ID is 0. Like the above, I think this should be `sort_order.order_id`.

File: python/pyiceberg/catalog/hive.py [hunk 14]
```diff
@@ -229,24 +290,43 @@ def create_table(
             ValueError: If the identifier is invalid
         """
         database_name, table_name = self.identifier_to_database_and_table(identifier)
-        current_time_millis = int(time.time())
+        seconds_since_epoch = int(time.time())
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:293]: The table metadata requires millis, but this produces seconds. Can you change this to set the table metadata correctly?

```python
current_time_millis = int(time.time() * 1000)
```

File: python/pyiceberg/catalog/hive.py [hunk 15]
```diff
@@ -229,24 +290,43 @@ def create_table(
...
+        metadata = TableMetadataV2(
+            location=location,
+            schemas=[schema],
+            current_schema_id=schema.schema_id,
+            partition_specs=[partition_spec],
+            default_spec_id=0,
+            sort_orders=[sort_order],
+            default_sort_order_id=0,
+            properties=properties or {},
+            last_updated_ms=seconds_since_epoch,
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:307]: Needs to be millis.

File: python/pyiceberg/catalog/hive.py [hunk 16]
```diff
@@ -229,24 +290,43 @@ def create_table(
...
+            last_column_id=schema.find_last_field_id(),
+            last_partition_id=partition_spec.spec_id,
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:309]: This needs to be the highest partition field ID, not the spec ID.

File: python/pyiceberg/schema.py [hunk 17]
```diff
@@ -169,6 +169,9 @@ def find_type(self, name_or_id: Union[str, int], case_sensitive: bool = True) ->
             raise ValueError(f"Could not find field with name or id {name_or_id}, case_sensitive={case_sensitive}")
         return field.field_type
 
+    def find_last_field_id(self):
```
Comment (rdblue) [inline python/pyiceberg/schema.py:172]: In Java, this is `highestFieldId`. Can we do the same here?

```
@property
def highest_field_id(self):
    ...
```

File: python/pyiceberg/serializers.py [hunk 18]
```diff
@@ -54,7 +54,13 @@ def table_metadata(input_file: InputFile, encoding: str = "utf-8") -> TableMetad
             TableMetadata: A table metadata instance
 
         """
-        return FromByteStream.table_metadata(byte_stream=input_file.open(), encoding=encoding)
+        # 😢 https://github.com/apache/iceberg/pull/5436
```
Comment (rdblue) [inline python/pyiceberg/serializers.py:57]: Merged.

File: python/pyiceberg/table/metadata.py [hunk 19]
```diff
@@ -55,7 +56,6 @@ def check_schemas(values: Dict[str, Any]) -> Dict[str, Any]:
 def check_partition_specs(values: Dict[str, Any]) -> Dict[str, Any]:
     """Validator to check if the default-spec-id is present in partition-specs"""
     default_spec_id = values["default_spec_id"]
-
```
Comment (rdblue) [inline python/pyiceberg/table/metadata.py:58]: Nit: Unnecessary whitespace change.

File: python/pyiceberg/catalog/hive.py (hunk 2 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:185]: Constructing a Hive schema from a schema, but I agree that a convertor is more appropriate.

File: python/pyiceberg/catalog/hive.py (hunk 3 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:169]: Good one, copied this one from legacy 👀

File: python/pyiceberg/catalog/hive.py (hunk 4 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:193]: Ahh, great catch! It was writing fine, but when reading the table it would throw an error:
```
Error: Error while processing statement: FAILED: Execution Error, return code 1 from org.apache.hadoop.hive.ql.exec.DDLTask. Error: type expected at the position 83 of 'string:int:boolean:array<string>:map<string,map<string,int>>:array<struct<latitude: float,longitude: float>>:struct<name: string,age: int>' but ' float' is found. (state=08S01,code=1)
```
Blame it on validate on read.

After removing all the spaces, it works fine:
```
+-------------------------------+----------------------------------------------------+----------------------------------------------------+
|           col_name            |                     data_type                      |                      comment                       |
+-------------------------------+----------------------------------------------------+----------------------------------------------------+
| # col_name                    | data_type                                          | comment                                            |
| foo                           | string                                             |                                                    |
| bar                           | int                                                |                                                    |
| baz                           | boolean                                            |                                                    |
| qux                           | array<string>                                      |                                                    |
| quux                          | map<string,map<string,int>>                        |                                                    |
| location                      | array<struct<latitude:float,longitude:float>>      |                                                    |
| person                        | struct<name:string,age:int>                        |                                                    |
|                               | NULL                                               | NULL                                               |
| # Detailed Table Information  | NULL                                               | NULL                                               |
| Database:                     | nyc                                                | NULL                                               |
| OwnerType:                    | USER                                               | NULL                                               |
| Owner:                        | fokkodriesprong                                    | NULL                                               |
| CreateTime:                   | Mon Aug 08 11:58:50 UTC 2022                       | NULL                                               |
| LastAccessTime:               | Mon Aug 08 11:58:50 UTC 2022                       | NULL                                               |
| Retention:                    | 0                                                  | NULL                                               |
| Location:                     | file:/Users/fokkodriesprong/Desktop/docker-spark-iceberg/wh/nyc.db/complex | NULL                                               |
| Table Type:                   | EXTERNAL_TABLE                                     | NULL                                               |
| Table Parameters:             | NULL                                               | NULL                                               |
|                               | EXTERNAL                                           | TRUE                                               |
|                               | metadata_location                                  | file:/Users/fokkodriesprong/Desktop/docker-spark-iceberg/wh/nyc.db/complex/metadata/3400b5e4-d454-4788-b269-18ea6a4beac0.metadata.json |
|                               | numFiles                                           | 0                                                  |
|                               | table_type                                         | ICEBERG                                            |
|                               | totalSize                                          | 0                                                  |
|                               | transient_lastDdlTime                              | 1659959930                                         |
|                               | NULL                                               | NULL                                               |
| # Storage Information         | NULL                                               | NULL                                               |
| SerDe Library:                | org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe | NULL                                               |
| InputFormat:                  | org.apache.hadoop.mapred.FileInputFormat           | NULL                                               |
| OutputFormat:                 | org.apache.hadoop.mapred.FileOutputFormat          | NULL                                               |
| Compressed:                   | No                                                 | NULL                                               |
| Num Buckets:                  | 0                                                  | NULL                                               |
| Bucket Columns:               | []                                                 | NULL                                               |
| Sort Columns:                 | []                                                 | NULL                                               |
+-------------------------------+----------------------------------------------------+----------------------------------------------------+
```

File: python/pyiceberg/catalog/hive.py (hunk 5 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:239]: I've made the value case-insensitive. Let me know if you also want to do the key.

File: python/pyiceberg/catalog/hive.py (hunk 6 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:240]: Sure thing. I've also split the exception into two parts; when the property `table_type` is missing, and when it is not equal to `iceberg`.

File: python/pyiceberg/catalog/hive.py (hunk 1 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:204]: I copied this from the docs, but they are very inconsistent. I tested, and the uppercase one actually fails when reading.

File: python/pyiceberg/schema.py (hunk 17 above)
Comment (Fokko) [reply python/pyiceberg/schema.py:172]: Sure thing, updated 👍🏻

File: python/pyiceberg/catalog/hive.py (hunk 16 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:309]: Auch, thanks!

File: python/pyiceberg/catalog/hive.py (hunk 15 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:307]: Great catch, Hive's `createTime` and `lastAccessTime` are in seconds

File: python/pyiceberg/catalog/hive.py (hunk 14 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:293]: Thanks, this was because Hive needs seconds.

File: python/pyiceberg/catalog/hive.py (hunk 13 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:305]: Great catch, thanks!

File: python/pyiceberg/catalog/hive.py (hunk 12 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:300]: Good one, we do have validators on them when we initialize the Pydantic models, but there is no re-assignment being done. Do you want to set them to zero? We could also set the `schema_id` by default to zero?

File: python/pyiceberg/catalog/hive.py (hunk 11 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:303]: Much better indeed

File: python/pyiceberg/catalog/hive.py (hunk 10 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:297]: I wasn't awe of that convention, but make sure to apply it consistently.

File: python/pyiceberg/catalog/hive.py (hunk 7 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:248]: I didn't want to spend too much time on it in this PR. I've created a stub for it 👍🏻

File: python/pyiceberg/catalog/hive.py (hunk 5 above)
Comment (rdblue) [reply python/pyiceberg/catalog/hive.py]: No need to alter the key. Just the value is fine.

File: python/pyiceberg/catalog/hive.py (hunk 12 above)
Comment (rdblue) [reply python/pyiceberg/catalog/hive.py:304]: Here's what we do in Java: https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/TableMetadata.java#L88-L133

When the table metadata builder adds a schema, spec, or order, the builder will check whether it is an existing order and find out the ID to assign. We should do something similar here.

This probably isn't something to do in this PR (since this is focused on Hive) but introducing the builder would probably be a good idea. That's how we also accumulate the change set that we pass to the REST catalog when committing table changes. I'd say let's move forward with what you have here (except passing the IDs rather than hard-coding to 0) and we can add reassignment later.

File: python/pyiceberg/catalog/hive.py (hunk 9 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:273]: Nice, added it, including a test

File: python/pyiceberg/catalog/hive.py (hunk 12 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:309]: Thanks for pointing out the Java code. I like it and agree that it is best to do it in separate PR.

File: python/pyiceberg/catalog/hive.py [hunk 20]
```diff
@@ -201,17 +238,46 @@ def __init__(self, name: str, properties: Properties, uri: str):
...
+            if prop_metadata_location := properties.get(METADATA_LOCATION):
+                metadata_location = prop_metadata_location
+            else:
+                raise NoSuchTableError("Metadata location not found")
+
+        file = io.new_input(metadata_location)
+        metadata = FromInputFile.table_metadata(file)
+        return Table(identifier=(table.dbName, table.tableName), metadata=metadata, metadata_location=metadata_location)
+
+    def _write_metadata(self, metadata: TableMetadataV2, io: FileIO, metadata_path: str):
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:260]: Why does this accept metadata v2 instead of either v1 or v2? I don't think that we want to assume that we will only write v2 metadata. We can't upgrade tables automatically.

File: python/pyiceberg/catalog/hive.py [hunk 21]
```diff
@@ -201,17 +238,46 @@ def __init__(self, name: str, properties: Properties, uri: str):
...
+        ToOutputFile.table_metadata(metadata, io.new_output(metadata_path))
+
+    def _resolve_table_location(self, location: Optional[str], database_name: str, table_name: str):
+        if not location:
+            if warehouse_location := self.properties.get(WAREHOUSE):
+                warehouse_location = warehouse_location.rstrip("/")
+                return f"{warehouse_location}/{database_name}/{table_name}"
+
+            database_properties = self.load_namespace_properties(database_name)
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:269]: The database location takes precedence over warehouse location.

File: python/pyiceberg/catalog/hive.py [hunk 22]
```diff
@@ -201,17 +238,46 @@ def __init__(self, name: str, properties: Properties, uri: str):
...
+            if database_location := database_properties.get(LOCATION):
+                database_location = database_location.rstrip("/")
+                return f"{database_location}/{table_name}"
+        raise ValueError("Cannot determine location from warehouse, please provide an explicit location")
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:273]: I think indentation for this is wrong. It should be inside the `if not location` part. At this level, we need a `return location`.

File: python/pyiceberg/catalog/hive.py [hunk 23]
```diff
@@ -233,24 +299,47 @@ def create_table(
             ValueError: If the identifier is invalid
         """
         database_name, table_name = self.identifier_to_database_and_table(identifier)
-        current_time_millis = int(time.time())
+        seconds_since_epoch = int(time.time() * 1000)
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:302]: This is in milliseconds, not seconds. I think you probably want to rename back to `current_time_millis`.

File: python/pyiceberg/catalog/hive.py [hunk 24]
```diff
@@ -233,24 +299,47 @@ def create_table(
...
+
+        location = self._resolve_table_location(location, database_name, table_name)
+
+        metadata_location = f"{location}/metadata/{uuid.uuid4()}.metadata.json"
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:306]: In Java, we add a version to the start of the filename so that we can tell the order. Can you do that here as well?

File: python/pyiceberg/catalog/hive.py [hunk 25]
```diff
@@ -233,24 +299,47 @@ def create_table(
...
+            sort_orders=[sort_order],
+            default_sort_order_id=sort_order.order_id,
+            properties=properties or {},
+            last_updated_ms=seconds_since_epoch,
+            last_column_id=schema.highest_field_id,
+            last_partition_id=max(field.field_id for field in partition_spec.fields)
+            if partition_spec.fields
+            else DEFAULT_LAST_PARTITION_ID,
+        )
+        io = load_file_io({**self.properties})
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:322]: If we're always using `self.properties` then I think it makes sense to have an `io` instance for the catalog. That can be used here rather than creating a new one per table.

File: python/pyiceberg/catalog/hive.py [hunk 26]
```diff
@@ -201,17 +238,46 @@ def __init__(self, name: str, properties: Properties, uri: str):
...
+
+        table_type = properties[TABLE_TYPE]
+        if table_type.lower() != ICEBERG:
+            raise NoSuchTableError(f"Property table_type is {table_type}, expected {ICEBERG}: {table.dbName}.{table.tableName}")
+
+        if not metadata_location:
+            if prop_metadata_location := properties.get(METADATA_LOCATION):
+                metadata_location = prop_metadata_location
+            else:
+                raise NoSuchTableError("Metadata location not found")
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:254]: It isn't that the file wasn't found, the problem is that `metadata_location` was missing from table properties.

File: python/pyiceberg/catalog/hive.py [hunk 27]
```diff
@@ -233,24 +299,47 @@ def create_table(
...
         )
         try:
             with self._client as open_client:
                 open_client.create_table(tbl)
                 hive_table = open_client.get_table(dbname=database_name, tbl_name=table_name)
         except AlreadyExistsException as e:
             raise TableAlreadyExistsError(f"Table {database_name}.{table_name} already exists") from e
-        return self._convert_hive_into_iceberg(hive_table)
+
+        return self._convert_hive_into_iceberg(hive_table, io, metadata_location)
```
Comment (rdblue) [inline python/pyiceberg/catalog/hive.py:342]: Why pass `metadata_location` here? Since it is set on the table, I think it would make sense to always pull it from table parameters. That way we never create a situation where we've forgotten to set the table parameter, but successfully returned a table instance.

File: python/pyiceberg/catalog/hive.py (hunk 20 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:260]: I've updated the signature. This is because we don't have any logic that allows updating a table. This is only used for creating new tables, and I think it makes sense to only allow V2 tables (or at least push the user into that direction).

File: python/pyiceberg/catalog/hive.py (hunk 21 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:269]: Thanks!

File: python/pyiceberg/catalog/hive.py (hunk 23 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:302]: Ugh, what a mess, thanks!

File: python/pyiceberg/catalog/hive.py (hunk 25 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:322]: I actually think that we should mix in the `properties` from the table in there as well. This way you can override this on a table level.

File: python/pyiceberg/catalog/hive.py (hunk 27 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:342]: I like that a lot. I think that the `metadata_location` was still there for historical reasons. Thanks!

File: python/pyiceberg/catalog/hive.py (hunk 24 above)
Comment (Fokko) [reply python/pyiceberg/catalog/hive.py:306]: Great suggestion, it is also in the spec. Added 👍🏻

File: python/tests/catalog/test_hive.py [hunk 28]
```diff
@@ -207,15 +250,137 @@ def test_create_table(table_schema_simple: Schema, hive_table: HiveTable):
...
     catalog = HiveCatalog(HIVE_CATALOG_NAME, {}, uri=HIVE_METASTORE_FAKE_URL)
 
     catalog._client = MagicMock()
     catalog._client.__enter__().get_table.return_value = hive_table
-    catalog.load_table(("default", "table"))
-
-    catalog._client.__enter__().get_table.assert_called_with(dbname="default", tbl_name="table")
+    table = catalog.load_table(("default", "new_tabl2e"))
+
+    catalog._client.__enter__().get_table.assert_called_with(dbname="default", tbl_name="new_tabl2e")
```
Comment (rdblue) [inline python/tests/catalog/test_hive.py:300]: Did you intend to have `2` in the table name?

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
@Fokko, looks like this is ready except that it conflicts with the order by PR that was just merged. I tried to fix it, but I didn't get it right.

**Comment by Fokko** [comment]
Awesome @rdblue just resolved the pre-commit error 👍🏻

**Comment by rdblue** [comment]
Thanks, @Fokko! Great to have this working.
